File: behavioral_pattern/strategy/CreditCardPayment.java
package strategy;

public class CreditCardPayment implements PaymentStrategy {
    private String cardNumber;
    private String name;

    public CreditCardPayment(String cardNumber, String name) {
        this.cardNumber = cardNumber;
        this.name = name;
    }

    @Override
    public void pay(int amount) {
        System.out.println("Paid $" + amount + " using Credit Card");
        System.out.println("Card Number: " + maskCardNumber(cardNumber));
        System.out.println("Card Holder: " + name);
    }

    private String maskCardNumber(String cardNumber) {
        return "****-****-****-" + cardNumber.substring(cardNumber.length() - 4);
    }
}

File: typescript/behavioral_pattern/src/strategy/CreditCardPayment.ts
import PaymentStrategy from "./PaymentStrategy";

class CreditCardPayment implements PaymentStrategy {
    private cardNumber: string;
    private name: string;

    public constructor(cardNumber: string, name: string) {
        this.cardNumber = cardNumber;
        this.name = name;
    }

    public pay(amount: number): void {
        console.log(`Paid ${amount} using Credit Card`);
        console.log(`Card Number: ${this.maskCardNumber(this.cardNumber)}`);
        console.log(`Card Holder: ${this.name}`);
    }

    public maskCardNumber(cardNumber: string): string {
        return "****-****-****-" + cardNumber.substring(cardNumber.length - 4);
    }
}

export default CreditCardPayment;